File: Cpp/String/Conversion.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    //6.Z字形变换
    string convert(string s, int numRows) {
        if(numRows <= 1){
            return s;
        }
        vector<string> res(numRows,"");
        int flag = -1;
        int i = 0;
        for(char c:s){
            if(i==0 || i==numRows-1){
                flag = -flag;
            }
            res[i] += c;
            i += flag;
        }
        string tmp = "";
        for(auto r:res){
            tmp += r;
        }
        return tmp;
    }

    //7.整数反转
    int reverse(int x) {
        int res = 0;
        int sign = (x<0)?-1:1;
        while(x){
            int tmp = abs(x) % 10;
            if(res > INT_MAX / 10 || (res == INT_MAX / 10 && tmp > INT_MAX % 10)){
                return 0;
            }
            if(res < INT_MIN / 10 || (res == INT_MIN / 10 && tmp > -(INT16_MIN % 10))){
                return 0;
            }
            x = abs(x) / 10;
            res = res*10 + sign * tmp;
        }
        return res;
    }

    //8.字符串转换整数(atoi)
    int myAtoi(string s) {
        //去除前导空格
        int i = 0;
        while(i < s.size() && s[i] == ' '){
            i++;
        }
        if(i == s.size()){
            return 0;
        }

        //处理第一个符号
        int sign = 1;
        if(s[i] == '-'){
            sign = -1;
            i++;
        }
        else if(s[i] == '+'){
            sign = 1;
            i++;
        }
        
        //获取最终结果
        int res = 0;
        while(i < s.size()){
            if(s[i] > '9' || s[i] < '0'){
                break;
            }
            if(res > INT_MAX / 10 || (res == INT_MAX / 10 && s[i] - '0' > INT_MAX % 10)) {
                return INT_MAX;
            }
            if(res < INT_MIN / 10 || (res == INT_MIN / 10 && s[i] - '0' > -(INT_MIN % 10))){
                return INT_MIN;
            }
            res = res * 10 + sign * (s[i] - '0');
            i++;
        }
        return res;
    }

    //9.回文数
    bool isPalindrome(int x) {
        if(x < 0){
            return false;
        }
        int tmp = x;
        int res = 0;
        while(tmp){
            int remainder = tmp % 10;
            tmp /= 10;
            res = res * 10 + remainder;
        } 
        return res == x;
    
    }

    //12.整数转罗马数字
    string intToRoman(int num) {
        string strs[] = {"M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"};
        int nums[] = {1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1};
        string res = "";
        for(int i=0; i<sizeof(nums)/sizeof(nums[0]); i++){
            int cnt = num / nums[i];
            while(cnt){
                res += strs[i];
                cnt--;
            }
            num = num % nums[i];
        }
        return res;
    }

    //13.罗马数字转整数
    int romanToInt(string s) {
        int res = 0;
        unordered_map<char,int> m {
            {'I',1},{'V',5},{'X',10},{'L',50},{'C',100},{'D',500},{'M',1000}
        };
        for(int i = 0; i < s.size() - 1; i++){
            int num1 = m[s[i]];
            int num2 = m[s[i+1]];
            if(num1 >= num2){
                res += num1;
            }
            else{
                res -= num1;
            }
        }
        res += m[s[s.size()-1]];
        return res;
    }
};

int main(){
    Solution s = Solution();
    string res1 = s.convert("PAYPALISHIRING",3);
    cout << res1 <<endl;

    int res2 = s.myAtoi("42");
    cout << res2 << endl;

    string res3 = s.intToRoman(1994);
    cout << res3 << endl;

    int res4 = s.reverse(-123);
    cout << res4 << endl;

    int res5 = s.romanToInt("III");
    cout << res5 << endl;

    bool res6 = s.isPalindrome(1234567899);
    cout << res6 << endl;
}

File: Python/Str/Conversion.py
class Solution:
    #6.Z字形变换
    def convert(self, s: str, numRows: int) -> str:
        if numRows < 2:
            return s
        res = [""] * numRows
        resIdx = 0
        flag = -1
        for c in s:
            res[resIdx] += c
            if resIdx == 0 or resIdx == numRows-1:
                flag = -flag
            resIdx += flag
        tmp = ""
        for i in res:
            tmp += i
        return tmp

    #7.整数反转
    def reverse(self, x: int) -> int:
        res = 0
        if x<0: sign = -1
        else: sign = 1
        while(x):
            if res > (2**31 - 1) / 10 or res < -2**31 / 10:
                return 0
            tmp = abs(x) % 10
            x = abs(x) // 10
            res = res * 10 + sign*tmp
        return res

    #8.字符串转换整数(atoi)
    def myAtoi(self, s: str) -> int:
        def delFrontSpace(s):
            i = 0
            while i<len(s) and s[i] == " ":
                i+=1
            return s[i:]

        def extractNum(s):
            if not s:
                return 0
            sign = 1
            if s[0] == "-":
                sign = -1
                s = s[1:]
            elif s[0] == "+":
                sign = 1
                s = s[1:]
            num = 0
            for i in range(len(s)):
                if s[i].isdigit():
                    num = num*10+int(s[i])
                else:
                    break
            num *= sign
            if num > 2 ** 31 - 1:
                num = 2 ** 31 - 1
            elif num < -2 ** 31:
                num = -2 ** 31
            return num

        s = delFrontSpace(s)
        num = extractNum(s)

        return num

    #9.回文数
    def isPalindrome(self, x: int) -> bool:
        origin = x
        res = 0
        if x<0:
            return False
        else:
            while x:
                pop = x % 10
                x = x // 10
                res = res * 10 + pop
        return res == origin

    #12.整数转罗马数字
    def intToRoman(self, num: int) -> str:
        res = ''
        d = {'M':1000,'CM':900,'D':500,'CD':400,'C':100,'XC':90,
             'L':50,'XL':40,'X':10,'IX':9,'V':5,'IV':4,'I':1}
        for key in d.keys():
            cnt = num // d[key]
            if cnt != 0:
                res += key * cnt
            num = num % d[key]
        return res

    #13.罗马数字转整数
    def romanToInt(self, s: str) -> int:
        res = 0
        d = {'M':1000,'CM':900,'D':500,'CD':400,'C':100,'XC':90,
             'L':50,'XL':40,'X':10,'IX':9,'V':5,'IV':4,'I':1}
        i = 0
        while(i<len(s)):
            if s[i:i+2] in d:
                res += d[s[i:i+2]]
                i += 2
            else:
                res += d[s[i]]
                i += 1
        return res


s = Solution()
res = s.convert("PAYPALISHIRING",3)
print(res)

res = s.reverse(-321)
print(res)

num = s.myAtoi('')
print(num)

res = s.isPalindrome(121)
print(res)

res = s.intToRoman(1994)
print(res)

res = s.romanToInt("MCMXCIV")
print(res)